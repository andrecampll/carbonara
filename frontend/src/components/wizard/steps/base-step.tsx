import { Button, Typography } from '@mui/material'
import { motion, Variants } from 'framer-motion'
import { memo, ReactNode } from 'react'
import { useWizard } from 'react-use-wizard'

const variants: Variants = {
  enter: (direction: number) => {
    return {
      x: direction > 0 ? 800 : -800,
      opacity: 0,
    }
  },
  center: {
    zIndex: 1,
    x: 0,
    opacity: 1,
  },
  exit: (direction: number) => {
    return {
      zIndex: 0,
      x: direction < 0 ? 800 : -800,
      opacity: 0,
    }
  },
}

type Props = {
  children: ReactNode
  title: string
  description: string
  hideCancelButton?: boolean
  onNextClick?: () => void
  nextButtonFormId?: string
}

export const BaseStep = memo<Props>(
  ({
    children,
    title,
    description,
    hideCancelButton = false,
    nextButtonFormId,
    onNextClick,
  }) => {
    const { activeStep, previousStep } = useWizard()

    const previousStepIndex = { current: activeStep - 1 }

    return (
      <motion.div
        custom={activeStep - previousStepIndex.current}
        variants={variants}
        initial="enter"
        animate="center"
        exit="exit"
        transition={{
          type: 'spring',
          stiffness: 300,
          damping: 30,
        }}
      >
        <div className="flex flex-col gap-4">
          <header>
            <Typography variant="h5">{title}</Typography>
            <Typography fontWeight={400}>{description}</Typography>
          </header>
          {children}

          <footer
            className={`flex w-full items-center ${!hideCancelButton ? 'justify-between' : 'justify-end'}`}
          >
            {!hideCancelButton && (
              <Button onClick={previousStep} color="secondary">
                Back
              </Button>
            )}

            <Button
              variant="contained"
              color="secondary"
              type="submit"
              form={nextButtonFormId}
              onClick={onNextClick}
            >
              Next
            </Button>
          </footer>
        </div>
      </motion.div>
    )
  },
)

BaseStep.displayName = 'BaseStep'
