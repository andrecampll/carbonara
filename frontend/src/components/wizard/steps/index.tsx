'use client'

import { AnimatePresence } from 'framer-motion'
import { ReactNode } from 'react'
import { Wizard } from 'react-use-wizard'

type Props = {
  children: ReactNode
}

export function Steps({ children }: Props) {
  return (
    <Wizard wrapper={<AnimatePresence initial={false} mode="wait" />}>
      {children}
    </Wizard>
  )
}
