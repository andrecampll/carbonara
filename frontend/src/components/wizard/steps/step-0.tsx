import Image from 'next/image'
import { useWizard } from 'react-use-wizard'

import { BaseStep } from './base-step'

export function Step0() {
  const { nextStep } = useWizard()

  return (
    <BaseStep
      title="Lets Calculate Your Carbon Footprint"
      description="Follow the steps to calculate your carbon footprint. We will provide
      you with a detailed report at the end. Lets get started!"
      hideCancelButton
      onNextClick={nextStep}
    >
      <div className="my-4 flex flex-col items-center gap-4">
        <Image
          alt=""
          width={200}
          height={200}
          src="/images/carbon-light.webp"
        />
      </div>
    </BaseStep>
  )
}
