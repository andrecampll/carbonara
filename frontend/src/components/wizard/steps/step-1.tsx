import { zodResolver } from '@hookform/resolvers/zod'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { useWizard } from 'react-use-wizard'
import { z } from 'zod'

import { Input } from '@/components/ui/input'
import { useFootprintWizard } from '@/hooks/use-footprint-wizard'

import { BaseStep } from './base-step'

const footprintStep1Schema = z.object({
  electrictyUsage: z.string().min(1, {
    message: 'Electricity usage is required',
  }),
})

type FootprintWizardFormStep1Data = z.infer<typeof footprintStep1Schema>

export function Step1() {
  const { nextStep } = useWizard()
  const { updateWizard, wizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FootprintWizardFormStep1Data>({
    resolver: zodResolver(footprintStep1Schema),
    defaultValues: {
      electrictyUsage: wizard?.electrictyUsage?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep1Data) => {
      // handle submit
      updateWizard({
        electrictyUsage: Number(formData.electrictyUsage),
      })
      nextStep()
    },
    [nextStep, updateWizard],
  )

  return (
    <BaseStep
      title="Electricity Usage"
      description="Enter your monthly electricity usage in kilowatt hours (Kwh) to
      calculate your carbon footprint."
      nextButtonFormId="electricity-usage-form"
    >
      <form
        onSubmit={handleSubmit(onSubmit)}
        className="my-4 flex flex-col items-center gap-4"
        id="electricity-usage-form"
      >
        <Image alt="" width={300} height={300} src="/images/electricity.webp" />

        <Input
          label="Electricity Usage (Kwh/Month):"
          type="number"
          helperText={errors.electrictyUsage?.message}
          {...register('electrictyUsage')}
        />
      </form>
    </BaseStep>
  )
}
