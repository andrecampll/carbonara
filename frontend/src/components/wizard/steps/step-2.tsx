import { zodResolver } from '@hookform/resolvers/zod'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { useWizard } from 'react-use-wizard'
import { z } from 'zod'

import { Input } from '@/components/ui/input'
import { useFootprintWizard } from '@/hooks/use-footprint-wizard'

import { BaseStep } from './base-step'

const footprintStep2Schema = z.object({
  transportationUsage: z.string().min(1, {
    message: 'Transportation gasoline usage is required',
  }),
})

type FootprintWizardFormStep2Data = z.infer<typeof footprintStep2Schema>

export function Step2() {
  const { nextStep } = useWizard()
  const { updateWizard, wizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FootprintWizardFormStep2Data>({
    resolver: zodResolver(footprintStep2Schema),
    defaultValues: {
      transportationUsage: wizard?.transportationUsage?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep2Data) => {
      updateWizard({
        transportationUsage: Number(formData.transportationUsage),
      })
      nextStep()
    },
    [nextStep, updateWizard],
  )

  return (
    <BaseStep
      title="Transportation Gasoline Usage"
      description="Enter your monthly transportation gasoline usage in gallons to
      calculate your carbon footprint."
      nextButtonFormId="transportation-usage-form"
    >
      <form
        onSubmit={handleSubmit(onSubmit)}
        className="my-4 flex flex-col items-center gap-4"
        id="transportation-usage-form"
      >
        <Image alt="" width={300} height={300} src="/images/gas.png" />

        <Input
          label="Transportation Gasoline Usage (Gallons/Month):"
          type="number"
          helperText={errors.transportationUsage?.message}
          {...register('transportationUsage')}
        />
      </form>
    </BaseStep>
  )
}
