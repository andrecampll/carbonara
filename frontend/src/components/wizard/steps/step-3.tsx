import { zodResolver } from '@hookform/resolvers/zod'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { useWizard } from 'react-use-wizard'
import { z } from 'zod'

import { Input } from '@/components/ui/input'
import { useFootprintWizard } from '@/hooks/use-footprint-wizard'

import { BaseStep } from './base-step'

const footprintStep3Schema = z.object({
  shortFlights: z.string().min(1, {
    message: 'Short flights is required',
  }),
  mediumFlights: z.string().min(1, {
    message: 'Medium flights is required',
  }),
  longFlights: z.string().min(1, {
    message: 'Long flights is required',
  }),
})

type FootprintWizardFormStep3Data = z.infer<typeof footprintStep3Schema>

export function Step3() {
  const { nextStep } = useWizard()
  const { updateWizard, wizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FootprintWizardFormStep3Data>({
    resolver: zodResolver(footprintStep3Schema),
    defaultValues: {
      shortFlights: wizard?.shortFlights?.toString() ?? '',
      mediumFlights: wizard?.mediumFlights?.toString() ?? '',
      longFlights: wizard?.longFlights?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep3Data) => {
      // handle submit
      updateWizard({
        shortFlights: Number(formData.shortFlights),
        mediumFlights: Number(formData.mediumFlights),
        longFlights: Number(formData.longFlights),
      })
      nextStep()
    },
    [nextStep, updateWizard],
  )

  return (
    <BaseStep
      title="Airplane Flights"
      description="Enter how many short, medium, and long flights you take per year to
      calculate your carbon footprint."
      nextButtonFormId="airplane-flights-form"
    >
      <form
        onSubmit={handleSubmit(onSubmit)}
        className="my-4 flex flex-col items-center gap-4"
        id="airplane-flights-form"
      >
        <Image alt="" width={300} height={300} src="/images/airplane.webp" />

        <Input
          label="Short Flights (4 hours):"
          type="number"
          helperText={errors.shortFlights?.message}
          {...register('shortFlights')}
        />
        <Input
          label="Medium Flights (6 hours):"
          type="number"
          helperText={errors.mediumFlights?.message}
          {...register('mediumFlights')}
        />
        <Input
          label="Long Flights (8 hours):"
          type="number"
          helperText={errors.longFlights?.message}
          {...register('longFlights')}
        />
      </form>
    </BaseStep>
  )
}
