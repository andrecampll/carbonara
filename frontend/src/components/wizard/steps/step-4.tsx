import { zodResolver } from '@hookform/resolvers/zod'
import Image from 'next/image'
import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { z } from 'zod'

import { SearchInput } from '@/components/ui/search-input'
import { CalculateFootprintMutationInput } from '@/data/types/footprint'
import {
  FootprintWizard,
  useFootprintWizard,
} from '@/hooks/use-footprint-wizard'

import { BaseStep } from './base-step'

const footprintStep4Schema = z.object({
  dietaryChoice: z.string().min(1, {
    message: 'Dietary choice is required',
  }),
})

type FootprintWizardFormStep4Data = z.infer<typeof footprintStep4Schema>

type Props = {
  onFinalSubmit: (
    wizardData: CalculateFootprintMutationInput['input'],
  ) => Promise<void>
}

export function Step4({ onFinalSubmit }: Props) {
  const { updateWizard, wizard, clearWizard } = useFootprintWizard()

  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
  } = useForm<FootprintWizardFormStep4Data>({
    resolver: zodResolver(footprintStep4Schema),
    defaultValues: {
      dietaryChoice: wizard?.dietaryChoice?.toString() ?? '',
    },
  })

  const onSubmit = useCallback(
    async (formData: FootprintWizardFormStep4Data) => {
      const dietaryChoice =
        formData.dietaryChoice as FootprintWizard['dietaryChoice']

      updateWizard({
        dietaryChoice,
      })

      await onFinalSubmit({
        electricityUsageKWh: Number(wizard?.electrictyUsage),
        transportationUsageGallonsPerMonth: Number(wizard?.transportationUsage),
        flightsShortHaul: Number(wizard?.shortFlights),
        flightsMediumHaul: Number(wizard?.mediumFlights),
        flightsLongHaul: Number(wizard?.longFlights),
        dietaryChoice,
      })

      clearWizard()
    },
    [updateWizard, onFinalSubmit, wizard, clearWizard],
  )

  return (
    <BaseStep
      title="Dietary Choice"
      description="Select your dietary choice to calculate your carbon footprint."
      nextButtonFormId="dietary-choice-form"
    >
      <form
        onSubmit={handleSubmit(onSubmit)}
        className="my-4 flex flex-col items-center gap-4"
        id="dietary-choice-form"
      >
        <Image alt="" width={300} height={300} src="/images/meat.png" />

        <SearchInput
          label="Dietary Choice:"
          helperText={errors.dietaryChoice?.message}
          {...register('dietaryChoice')}
          onChangeValue={(value) => {
            setValue('dietaryChoice', value)
          }}
        />
      </form>
    </BaseStep>
  )
}
