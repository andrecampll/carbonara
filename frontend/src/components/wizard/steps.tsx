'use client'

import { useMutation } from '@apollo/client'
import { useCallback } from 'react'
import { Wizard } from 'react-use-wizard'

import {
  CalculateFootprintMutationInput,
  Footprint,
} from '@/data/types/footprint'
import { CALCULATE_FOOTPRINT_MUTATION } from '@/graphql/mutations/calculateFootprint'
import { useFootprint } from '@/hooks/use-footprint'

import { Step0 } from './step-0'
import { Step1 } from './step-1'
import { Step2 } from './step-2'
import { Step3 } from './step-3'
import { Step4 } from './step-4'

type CalculateFootprintMutationResponse = {
  calculate: {
    footprint: Footprint
  }
}

export function Steps() {
  const { addFootprint } = useFootprint()

  const [calculateFunction] = useMutation<CalculateFootprintMutationResponse>(
    CALCULATE_FOOTPRINT_MUTATION,
  )

  const handleCalculateFootprint = useCallback(
    async (wizardData: CalculateFootprintMutationInput['input']) => {
      const { data: footPrintData } = await calculateFunction({
        variables: {
          input: wizardData,
        },
      })

      addFootprint(footPrintData?.calculate.footprint as Footprint)
    },
    [calculateFunction, addFootprint],
  )

  return (
    <Wizard>
      <Step0 />
      <Step1 />
      <Step2 />
      <Step3 />
      <Step4 onFinalSubmit={handleCalculateFootprint} />
    </Wizard>
  )
}
