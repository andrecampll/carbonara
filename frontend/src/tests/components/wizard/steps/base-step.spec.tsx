import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react'
import { Wizard } from 'react-use-wizard'
import { afterEach, describe, expect, it } from 'vitest'

import { BaseStep } from '../../../../components/wizard/steps/base-step'

afterEach(cleanup)

describe('BaseStep', () => {
  it('should be able to render BaseStep component', () => {
    const { container } = render(
      <Wizard>
        <BaseStep
          title="Lets Calculate Your Carbon Footprint"
          description="Follow the steps to calculate your carbon footprint. We will provide
      you with a detailed report at the end. Lets get started!"
          onNextClick={() => {}}
        >
          <div>
            <p>Hello World</p>
          </div>
        </BaseStep>
      </Wizard>,
    )

    expect(screen.getByText(/Hello World/i)).toBeDefined()
    expect(
      screen.getByText(/Lets Calculate Your Carbon Footprint/i),
    ).toBeDefined()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should be able to render BaseStep component without cancel button', () => {
    const { container } = render(
      <Wizard>
        <BaseStep
          title="Lets Calculate Your Carbon Footprint"
          description="Follow the steps to calculate your carbon footprint. We will provide
      you with a detailed report at the end. Lets get started!"
          onNextClick={() => {}}
          hideCancelButton
        >
          <div>
            <p>Hello World</p>
          </div>
        </BaseStep>
      </Wizard>,
    )

    expect(screen.queryByRole('button', { name: /back/i })).toBeNull()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should be able to render Base component and go back', async () => {
    const { container } = render(
      <Wizard>
        <BaseStep title="BaseStep title" description="BaseStep description">
          <div>BaseStep test</div>
        </BaseStep>
      </Wizard>,
    )

    const backButton = screen.getByRole('button', { name: /back/i })

    fireEvent.click(backButton)

    await waitFor(
      () => {
        expect(screen.getByText(/BaseStep test/i)).toBeDefined()
      },
      { timeout: 1000 },
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
