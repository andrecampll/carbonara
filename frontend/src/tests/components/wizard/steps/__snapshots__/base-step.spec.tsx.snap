// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseStep > should be able to render Base component and go back 1`] = `
<div
  style="opacity: 0; transform: translateX(800px) translateZ(0); z-index: 1;"
>
  <div
    class="flex flex-col gap-4"
  >
    <header>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        BaseStep title
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3sis8q-MuiTypography-root"
      >
        BaseStep description
      </p>
    </header>
    <div>
      BaseStep test
    </div>
    <footer
      class="flex w-full items-center justify-between"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Back
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Next
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </footer>
  </div>
</div>
`;

exports[`BaseStep > should be able to render BaseStep component 1`] = `
<div
  style="opacity: 0; transform: translateX(800px) translateZ(0);"
>
  <div
    class="flex flex-col gap-4"
  >
    <header>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Lets Calculate Your Carbon Footprint
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3sis8q-MuiTypography-root"
      >
        Follow the steps to calculate your carbon footprint. We will provide
      you with a detailed report at the end. Lets get started!
      </p>
    </header>
    <div>
      <p>
        Hello World
      </p>
    </div>
    <footer
      class="flex w-full items-center justify-between"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Back
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Next
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </footer>
  </div>
</div>
`;

exports[`BaseStep > should be able to render BaseStep component without cancel button 1`] = `
<div
  style="opacity: 0; transform: translateX(800px) translateZ(0);"
>
  <div
    class="flex flex-col gap-4"
  >
    <header>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Lets Calculate Your Carbon Footprint
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3sis8q-MuiTypography-root"
      >
        Follow the steps to calculate your carbon footprint. We will provide
      you with a detailed report at the end. Lets get started!
      </p>
    </header>
    <div>
      <p>
        Hello World
      </p>
    </div>
    <footer
      class="flex w-full items-center justify-end"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Next
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </footer>
  </div>
</div>
`;
